### PR TITLE
Convert distributed tests to pytest

### PR DESCRIPTION
**Pull request by vepadulano**
This is an alternative to https://github.com/root-project/roottest/pull/838 aimed at improving the sporadic issues with timeouts happening in the CI nodes.

As pointed out in the other PR, it would be better to have only one connection object per backend test suite, which would then be shared by all tests of the same backend.

There are 2 major obstacles to that objective with the current status:
1. Python standard `unittest` support for fixtures is very limited. Namely, there is no way that I could find to have a singleton connection object shared among tests written in different files.
2. Tests in this repository are eventually run through `ctest`. This means that each test is run as its own separate `python test_file.py` call.

This PR addresses both issues with the following. To solve issue 1, All distributed RDataFrame tests are moved to `pytest`. Fixture support in this package is much more robust and for example allows sharing a fixture by all tests that are run in the same Python session. Unfortunately, issue 2 prevents us from running all the tests in separate files as they are now because they would not be sharing the same Python session. To overcome this, a single test file called `test_all.py` is created in each sub-directory corresponding to a different backend. This file then imports all other tests and will thus run them in the same python session when `ctest` will invoke it.

### CODE REVIEW

**Comment by vepadulano** [comment]
That was a bit anticlimactic :smile: let us try to understand a bit more.

In principle, running all tests in a single execution means bringing all the runtime into a single test. If previously the overall runtime on a node for a certain backend was higher than 5 minutes, it might well be that the new way of running tests incurs in a timeout by design.

On my own machine, I notice a tangible improvement (1-2 minutes to run the full `spark` folder compared to 7 minutes). Let's see on the CI nodes.

ubuntu2004:
* [latest nightly](https://lcgapp-services.cern.ch/root-jenkins/view/ROOT%20Nightly/job/root-nightly-master/LABEL=ROOT-ubuntu2004,SPEC=cxx17,V=master/lastBuild/testReport/projectroot.roottest.python.distrdf/): 7 minutes for dask, 8 for spark
* [this PR](https://lcgapp-services.cern.ch/root-jenkins/job/roottest-pullrequests-build/10654/testReport/projectroot.roottest.python.distrdf/): 3 min for both backends

mac 1015:
* [latest nightly with no failures in distrdf tests](https://lcgapp-services.cern.ch/root-jenkins/view/ROOT%20Nightly/job/root-nightly-master/LABEL=mac1015,SPEC=noimt,V=master/3121/testReport/projectroot.roottest.python.distrdf/): 13 min dask, 22 min spark

mac 11:
* [latest nightly with no failures in distrdf tests](https://lcgapp-services.cern.ch/root-jenkins/view/ROOT%20Nightly/job/root-nightly-master/LABEL=mac11,SPEC=default,V=master/3123/testReport/projectroot.roottest.python.distrdf/): 10 min for each backend

So it seems like the mac nodes are much slower. It becomes thus more plausible that we got the timeouts in this PR since they were set at "only" 5 minutes.

I could still make the distrdf tests run serially in any case, which is still not done at this point of the PR, but more probably I should increase the timeout for these tests

**Comment by eguiraud** [comment]
Mmh 35 minutes to run all distrdf tests seems crazy high, more than a factor 5 w.r.t. Ubuntu can't be just slow CPUs...? I seem to remember Macs were slow in resolving hostnames sometimes or something similar, could this be it? And: these are not the runtimes Mac users see, are they? :)

**Comment by vepadulano** [comment]
About the runtimes, I'm not even sure taking these tests is a fair measure, ctest + Python unittest already add some overhead by themselves. I don't really know how to test what a mac user would see as a generic runtime, what I can do is running the test suite manually on one of the mac nodes and see if it takes 35 minutes.

I'm not aware of the issue with hostnames, again I can only test this on the CI nodes manually. I'll run the tests and see if something is noticeably slow

**Comment by vepadulano** [comment]
Meanwhile, I ran the tests on `macitois21` which is mostly free at the moment:

```
macitois21:build sftnight$ ctest -R roottest-python-distrdf
Test project /Users/sftnight/build/night/LABEL/mac1015/SPEC/noimt/V/master/build
      Start 1435: roottest-python-distrdf-common-common_test_rungraphs
 1/19 Test #1435: roottest-python-distrdf-common-common_test_rungraphs .......   Passed   61.92 sec
      Start 1436: roottest-python-distrdf-dask-dask_test_backend
 2/19 Test #1436: roottest-python-distrdf-dask-dask_test_backend .............   Passed   35.19 sec
      Start 1437: roottest-python-distrdf-dask-dask_test_definepersample
 3/19 Test #1437: roottest-python-distrdf-dask-dask_test_definepersample .....   Passed   30.53 sec
      Start 1438: roottest-python-distrdf-dask-dask_test_friend_trees
 4/19 Test #1438: roottest-python-distrdf-dask-dask_test_friend_trees ........   Passed   32.27 sec
      Start 1439: roottest-python-distrdf-dask-dask_test_histo_write
 5/19 Test #1439: roottest-python-distrdf-dask-dask_test_histo_write .........   Passed   13.47 sec
      Start 1440: roottest-python-distrdf-dask-dask_test_include_headers
 6/19 Test #1440: roottest-python-distrdf-dask-dask_test_include_headers .....   Passed   16.52 sec
      Start 1441: roottest-python-distrdf-dask-dask_test_inv_mass
 7/19 Test #1441: roottest-python-distrdf-dask-dask_test_inv_mass ............   Passed   26.53 sec
      Start 1442: roottest-python-distrdf-dask-dask_test_rungraphs
 8/19 Test #1442: roottest-python-distrdf-dask-dask_test_rungraphs ...........   Passed   21.96 sec
      Start 1443: roottest-python-distrdf-dask-dask_test_reducer_merge
 9/19 Test #1443: roottest-python-distrdf-dask-dask_test_reducer_merge .......   Passed  108.56 sec
      Start 1444: roottest-python-distrdf-dask-dask_test_variations
10/19 Test #1444: roottest-python-distrdf-dask-dask_test_variations ..........   Passed   43.85 sec
      Start 1445: roottest-python-distrdf-spark-spark_test_backend
11/19 Test #1445: roottest-python-distrdf-spark-spark_test_backend ...........   Passed   35.52 sec
      Start 1446: roottest-python-distrdf-spark-spark_test_definepersample
12/19 Test #1446: roottest-python-distrdf-spark-spark_test_definepersample ...   Passed   33.99 sec
      Start 1447: roottest-python-distrdf-spark-spark_test_friend_trees
13/19 Test #1447: roottest-python-distrdf-spark-spark_test_friend_trees ......   Passed   30.81 sec
      Start 1448: roottest-python-distrdf-spark-spark_test_histo_write
14/19 Test #1448: roottest-python-distrdf-spark-spark_test_histo_write .......   Passed   16.52 sec
      Start 1449: roottest-python-distrdf-spark-spark_test_include_headers
15/19 Test #1449: roottest-python-distrdf-spark-spark_test_include_headers ...   Passed   20.05 sec
      Start 1450: roottest-python-distrdf-spark-spark_test_inv_mass
16/19 Test #1450: roottest-python-distrdf-spark-spark_test_inv_mass ..........   Passed   28.99 sec
      Start 1451: roottest-python-distrdf-spark-spark_test_rungraphs
17/19 Test #1451: roottest-python-distrdf-spark-spark_test_rungraphs .........   Passed   25.01 sec
      Start 1452: roottest-python-distrdf-spark-spark_test_reducer_merge
18/19 Test #1452: roottest-python-distrdf-spark-spark_test_reducer_merge .....   Passed  102.63 sec
      Start 1453: roottest-python-distrdf-spark-spark_test_variations
19/19 Test #1453: roottest-python-distrdf-spark-spark_test_variations ........   Passed   40.88 sec

100% tests passed, 0 tests failed out of 19

Total Test time (real) = 725.73 sec
```

So it's 5.5 minutes each backend, not too bad considering that each of those tests is creating and destroying the connection objects

**Comment by eguiraud** [comment]
> each of those tests is creating and destroying the connection objects

is that required? i thought the suggestion at https://github.com/root-project/roottest/pull/838#discussion_r816690259 was to reuse the schedulers once?

**Comment by vepadulano** [comment]
> is that required? i thought the suggestion at https://github.com/root-project/roottest/pull/838#discussion_r816690259 was to reuse the schedulers once?

The test run I show above is just using the latest nightly installation, so the old roottest running Python unittest. I am right now building a custom installation of ROOT on macitois21 with the new  pytest-based tests of this PR so we can compare the runtime difference

**Comment by eguiraud** [comment]
Ah sorry I misunderstood. Then what you meant is that things on `macitois21` are likely ok (pending confirmation on the new logic), so it's not a generic mac issue but a problem of the mac1015 and mac11 nodes?

That's good I guess, and we can point out which build nodes are not up to par at the next team meeting.

**Comment by vepadulano** [comment]
>  Then what you meant is that things on macitois21 are likely ok (pending confirmation on the new logic), so it's not a generic mac issue but a problem of the mac1015 and mac11 nodes?

Eh, not really unfortunately. The same machine in the latest nightly build has a timeout on the [dask_test_reducer_merge](https://lcgapp-services.cern.ch/root-jenkins/view/ROOT%20Nightly/job/root-nightly-master/LABEL=mac1015,SPEC=noimt,V=master/lastBuild/testReport/projectroot.roottest.python.distrdf/dask/roottest_python_distrdf_dask_dask_test_reducer_merge/) test and takes 15 minutes for each backend https://lcgapp-services.cern.ch/root-jenkins/view/ROOT%20Nightly/job/root-nightly-master/LABEL=mac1015,SPEC=noimt,V=master/lastBuild/testReport/projectroot.roottest.python.distrdf/ 

I think it's more a problem of how much load there is on the machine at any given time and/or some slowness in resolving hostnames or opening ports for the connections

**Comment by vepadulano** [comment]
I was finally able to test with the content of this PR on `macitois21`:
```
macitois21:master-distrdf-debug sftnight$ ctest -R roottest-python-distrdf
Test project /Users/sftnight/vpadulan/rootbuild/master-distrdf-debug
    Start 907: roottest-python-distrdf-common-test_all
1/3 Test #907: roottest-python-distrdf-common-test_all ...   Passed   59.17 sec
    Start 908: roottest-python-distrdf-dask-test_all
2/3 Test #908: roottest-python-distrdf-dask-test_all .....   Passed  232.56 sec
    Start 909: roottest-python-distrdf-spark-test_all
3/3 Test #909: roottest-python-distrdf-spark-test_all ....   Passed  199.55 sec

100% tests passed, 0 tests failed out of 3
```

As expected, the runtime decreases noticeably (albeit not as much as on my laptop or on the ubuntu CI node). Keep in mind that these are the runtimes taken right now, with `macitois` being quite free (`Load Avg: 2.73, 3.06, 3.00 CPU usage: 1.5%`). When a ROOT build runs and tests, the load is much higher (it's always close to 100% CPU usage). It's then almost guaranteed that those 232 seconds (dask) and 199 seconds (spark) written above easily become more than 300 seconds, thus triggering timeouts.

Bottom line: these tests are inherently long, this PR brings runtime down, but with the load that we impose on the nodes it's almost guaranteed that we always trigger the 5 minute timeout. IMO we should just increase the timeout in this PR

**Comment by vepadulano** [comment]
@pcanal ran the Dask tests on his machine (thanks!) and this is the report
```
Test project /Users/pcanal/root_working/root_builds/master.dist_rdf
    Start 898: roottest-python-distrdf-dask-dask_test_backend
1/9 Test #898: roottest-python-distrdf-dask-dask_test_backend ...........   Passed   19.38 sec
    Start 899: roottest-python-distrdf-dask-dask_test_definepersample
2/9 Test #899: roottest-python-distrdf-dask-dask_test_definepersample ...   Passed   17.30 sec
    Start 900: roottest-python-distrdf-dask-dask_test_friend_trees
3/9 Test #900: roottest-python-distrdf-dask-dask_test_friend_trees ......   Passed   17.58 sec
    Start 901: roottest-python-distrdf-dask-dask_test_histo_write
4/9 Test #901: roottest-python-distrdf-dask-dask_test_histo_write .......   Passed    7.33 sec
    Start 902: roottest-python-distrdf-dask-dask_test_include_headers
5/9 Test #902: roottest-python-distrdf-dask-dask_test_include_headers ...   Passed    9.34 sec
    Start 903: roottest-python-distrdf-dask-dask_test_inv_mass
6/9 Test #903: roottest-python-distrdf-dask-dask_test_inv_mass ..........   Passed   37.80 sec
    Start 904: roottest-python-distrdf-dask-dask_test_rungraphs
7/9 Test #904: roottest-python-distrdf-dask-dask_test_rungraphs .........   Passed   11.84 sec
    Start 905: roottest-python-distrdf-dask-dask_test_reducer_merge
8/9 Test #905: roottest-python-distrdf-dask-dask_test_reducer_merge .....   Passed   57.18 sec
```

So roughly 4x faster than what we see in the nightlies / PR builds on the mac 10.15 nodes.

**Comment by vepadulano** [comment]
Runtimes in the latest CI run:

* [mac1015](https://lcgapp-services.cern.ch/root-jenkins/job/roottest-pullrequests-build/10669/testReport/projectroot.roottest.python.distrdf/): 3m47s dask, 3m20s spark
* [mac11](https://lcgapp-services.cern.ch/root-jenkins/job/roottest-pullrequests-build/10670/testReport/projectroot.roottest.python.distrdf/): 2m22s dask, 2m03s spark
* [ubuntu2004](https://lcgapp-services.cern.ch/root-jenkins/job/roottest-pullrequests-build/10670/testReport/projectroot.roottest.python.distrdf/): 2m01s dask, 1m55s spark

**Comment by vepadulano** [comment]
Latest CI results:
* [mac1015](https://lcgapp-services.cern.ch/root-jenkins/job/roottest-pullrequests-build/10711/testReport/projectroot.roottest.python.distrdf/): dask 3 min 57 sec, spark 3 min 57 sec
* [mac11](https://lcgapp-services.cern.ch/root-jenkins/job/roottest-pullrequests-build/10712/testReport/projectroot.roottest.python.distrdf/): dask 4 min 24 sec, spark 3 min 49 sec
* [ubuntu2004](https://lcgapp-services.cern.ch/root-jenkins/job/roottest-pullrequests-build/10710/testReport/projectroot.roottest.python.distrdf/): dask 2 min 8 sec, spark 2 min 3 sec

This seems promising, given the tests on mac already run for 4 minutes probably the timeout should still be raised to more than 5 minutes, but the tests can run in parallel with the others thanks to the `PROCESSORS` test property.

I'll add a way to pass the property directly to the `ADD_TEST` function then come back to this PR

**Review by eguiraud — APPROVED** [review]
LGTM (shallow review)

**Review by etejedor — APPROVED** [review]
LGTM! As discussed, just a few comments would perhaps improve the PR.
